Astronomy/View Controllers/PhotosCollectionViewController.swift

```diff
@@ -64,13 +71,37 @@ class PhotosCollectionViewController: UIViewController, UICollectionViewDataSour
     
     private func loadImage(forCell cell: ImageCollectionViewCell, forItemAt indexPath: IndexPath) {
         
-        // let photoReference = photoReferences[indexPath.item]
+        let photoReference = photoReferences[indexPath.item]
+        let imageURL = photoReference.imageURL.usingHTTPS!
+        
+        if let cachedImage = cache.value(for: photoReference.id) {
+            cell.imageView.image = cachedImage
+            print("Images were set from cache")
+            return
+        }
         
         // TODO: Implement image loading here
+        
+        let task = URLSession.shared.dataTask(with: imageURL) { (data, _, error) in
```

Comment:
You never actually used your FetchPhotoOperation...

 

> Update PhotosCollectionViewController to Use Operations
> 
> Now that you've created an operation to fetch images, you should rewrite PhotosCollectionViewController so that it uses operations to do image loading.
> 
> Add a private property called photoFetchQueue, which is an instance of OperationQueue.
> In PhotosCollectionViewController.loadImage(forCell:, forItemAt:), delete the code that creates a data task.
> Create three operations:
> One should be a PhotoFetchOperation to fetch the image data from the network.
> One should be used to store received data in the cache.
> The last should check if the cell has been reused, and if not, set its image view's image.
> The last two of these can be instances of BlockOperation.
> Make the cache and completion operations both depend on completion of the fetch operation.
> Add each operation to the appropriate queue. Note that the last operation above uses UIKit API and must run on the main queue.
> Add a dictionary property that you'll use to store fetch operations by the associated photo reference id.
> When you finish creating and starting the operations for a cell, add the fetch operation to your dictionary. This way you can retrieve it later to cancel it if need be.
> Test
> 
> Your app should behave as it did before, loading images and scrolling. Because you haven't yet implemented cancellation, the problem with lower-down images loading slowly is not yet fixed.
> 
> Implement Cancellation
> 
> The last piece of the puzzle is to implement cancellation of operations that are in-flight but not yet complete when the associated cell scrolls ofscreen.
> 
> Implement the UICollectionViewDelegate method collectionView(_:, didEndDisplaying:, forItemAt:), which is called when a given item scrolls off screen.
> In your implementation, get the associated fetch operation and cancel it.
> Test One Last Time
> 
> At this point, your application should perform well even on a (simulated) slow network. You should be able to scroll all the way to the bottom of a large page of images and see the bottom-most images load relatively quickly without waiting for all previous images to load first. You can add print statements to verify that operations are being cancelled as they should.